### PR TITLE
Get template

### PR DESCRIPTION
**Pull request by gudgl**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A tested in insomnia
- [ ] Test B storage tests added action tests will be added in templates-servise-refactor branch

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: storage/s3/templates.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const bucket = "BUCKET HERE"
```
Comment (FilipNikolovski) [inline storage/s3/templates.go:11]: os.Getenv("TEMPLATES_BUCKET")

Comment (gudgl) [reply storage/s3/templates.go:11]: This whole file is removed with latest changes with the service.

File: services/templates/service.go [hunk 2]
```diff
@@ -62,23 +78,110 @@ func (s service) AddTemplate(c context.Context, template *entities.Template) err
...
 // GetTemplate returns the template with given template id and user id
 func (s service) GetTemplate(c context.Context, templateID int64, userID int64) (*entities.Template, error) {
-	template, err := storage.GetTemplate(c, templateID, userID)
+	template, err := s.db.GetTemplate(templateID, userID)
 	if err != nil {
 		return nil, fmt.Errorf("get template: %w", err)
 	}
 
-	html, err := s3.GetHTMLTemplate(c, userID, template.Name)
+	obj, err := s.s3.GetObject(&s3.GetObjectInput{
```
Comment (FilipNikolovski) [inline services/templates/service.go:124]: Check out the example for the method here: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.GetObject

```golang
svc := s3.New(session.New())
input := &s3.GetObjectInput{
    Bucket: aws.String("examplebucket"),
    Key:    aws.String("HappyFace.jpg"),
}

result, err := svc.GetObject(input)
if err != nil {
    if aerr, ok := err.(awserr.Error); ok {
        switch aerr.Code() {
        case s3.ErrCodeNoSuchKey:
            fmt.Println(s3.ErrCodeNoSuchKey, aerr.Error())
        case s3.ErrCodeInvalidObjectState:
            fmt.Println(s3.ErrCodeInvalidObjectState, aerr.Error())
        default:
            fmt.Println(aerr.Error())
        }
    } else {
        // Print the error, cast err to awserr.Error to get the Code and
        // Message from an error.
        fmt.Println(err.Error())
    }
    return
}

fmt.Println(result)
```

We can have a finer-grained error messages to properly log them maybe. Also, do we need to close the body here?

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

File: services/templates/service.go (hunk 2 above)
Comment (gudgl) [reply services/templates/service.go]: @FilipNikolovski I made some changes you can review them

File: services/templates/service.go [hunk 3]
```diff
@@ -13,3 +19,20 @@ type service struct {
...
 }
+
+// GetTemplate returns the template with given template id and user id
+func (s service) GetTemplate(c context.Context, templateID int64, userID int64) (*entities.Template, error) {
+	template, err := storage.GetTemplate(c, templateID, userID)
+	if err != nil {
+		return nil, fmt.Errorf("get template: %w", err)
+	}
+
+	html, err := s3.GetHTMLTemplate(c, userID, template.Name)
```
[the remark this answers is not in the document]
Comment (gudgl) [reply services/templates/service.go]: @FilipNikolovski i can not test this i still dont have access fot `GetObject`

File: services/templates/service.go [hunk 4]
```diff
@@ -121,16 +125,26 @@ func (s service) GetTemplate(c context.Context, templateID int64, userID int64)
...
+			case s3.ErrCodeNoSuchKey:
+				return nil, ErrHTMLPartNotFound
+			case s3.ErrCodeInvalidObjectState:
+				return nil, ErrInvalidHTMLPart
+			default:
+				return nil, fmt.Errorf("get object: %w", aerr)
+			}
+		}
 		return nil, fmt.Errorf("get object: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (gudgl) [reply services/templates/service.go:172]: yes you are right

File: services/templates/service.go [hunk 5]
```diff
@@ -144,8 +170,10 @@ func (s service) GetTemplate(c context.Context, templateID int64, userID int64)
...
 
+	defer resp.Body.Close()
+
 	var html []byte
-	_, err = result.Body.Read(html)
+	_, err = resp.Body.Read(html)
```
Comment (FilipNikolovski) [inline services/templates/service.go:176]: We should utilize `ioutil.ReadAll` here I believe:

```go
htmlBytes, err := ioutil.ReadAll(resp.Body)
```

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

File: services/templates/service.go (hunk 5 above)
Comment (gudgl) [reply services/templates/service.go:176]: done

File: actions/templates.go [hunk 6]
```diff
@@ -85,7 +84,7 @@ func GetTemplates(c *gin.Context) {
 
 	scopeMap := c.QueryMap("scopes")
 
-	s := templatesvc.NewTemplateService(storage.GetFromContext(c), s3.GetFromContext(c))
+	s := templatesvc.New(storage.GetFromContext(c), s3.GetFromContext(c))
```
Comment (FilipNikolovski) [inline actions/templates.go:87]:
```suggestion
	service := templatesvc.New(storage.GetFromContext(c), s3.GetFromContext(c))
```

Be consistent in naming of the variables throughout the code.

File: services/templates/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gudgl) [reply services/templates/service.go]: @FilipNikolovski i tested it with `errgroup` and it seemd a little bit slower so i will leave it as it is without `errgroup`

File: services/templates/service.go [hunk 7]
```diff
@@ -141,3 +144,87 @@ func (s *service) DeleteTemplate(c context.Context, templateID, userID int64) er
...
+	if err != nil {
+		return nil, fmt.Errorf("read: %w", err)
+	}
+
+	template.HTMLPart = string(html)
+
+	return template, nil
+}
+
+/*func (s service) getTemplate(c context.Context, templateID int64, userID int64) (*entities.Template, error) {
```
Comment (FilipNikolovski) [inline services/templates/service.go:186]: @gudgl let's test and implement this function with the errgroup. From what I can tell I think it's mostly correct.

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

File: services/templates/service.go (hunk 7 above)
Comment (gudgl) [reply services/templates/service.go]: @FilipNikolovski i am resolving this since i commented on the previous comment about the same purpose

**Comment by FilipNikolovski** [comment]
@gudgl the code review fails check it here: https://github.com/mailbadger/app/pull/934/checks?check_run_id=1627822086
It's missing an err check for the `defer Body.Close()`.

Check the [importer.go line](https://github.com/mailbadger/app/blob/master/services/subscribers/importer/importer.go#L54) to see how it's implemented elsewhere.

**Review by FilipNikolovski — APPROVED** [review]

**Pull request merged**